File: main2.py
import cv2
import time
from deepface import DeepFace

def detect_faces(frame, face_cascade):
    scaleFactor = 1.05
    minNeighbors = 20
    faces = face_cascade.detectMultiScale(frame, scaleFactor, minNeighbors)
    return faces

def detect_expression(frame, face_cascade):
    faces = detect_faces(frame, face_cascade)
    emotions = []
    if len(faces) == 0:
        return "No face detected"
    
    for (x, y, w, h) in faces:
        face_frame = frame[y:y+h, x:x+w]
        try:
            analysis_results = DeepFace.analyze(face_frame, actions=['emotion'], enforce_detection=False)
            
            if isinstance(analysis_results, list) and len(analysis_results) > 0:
                analysis = analysis_results[0]
                emotion_scores = analysis['emotion']
                dominant_emotion = max(emotion_scores, key=emotion_scores.get)
                emotions.append(dominant_emotion)
            else:
                emotions.append("Emotion data not found")

        except Exception as e:
            print(f"Error in emotion detection: {e}")
            emotions.append("Error in detection")

    return ", ".join(emotions) if emotions else "No face/emotion detected"

def main():
    face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')
    cap = cv2.VideoCapture(0)

    last_analysis_time = time.time()

    while True:
        ret, frame = cap.read()
        if not ret:
            break

        if time.time() - last_analysis_time >= 5:
            last_analysis_time = time.time()
            expression = detect_expression(frame, face_cascade)
            print("Dominant Emotion:", expression)

        cv2.imshow('Video', frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    cap.release()
    cv2.destroyAllWindows()

if __name__ == "__main__":
    main()

File: main2.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <thread>
#include <atomic>
#include <chrono>

std::atomic<bool> finished(false);
std::atomic<bool> playAlert(false);

void playSound(const std::string& soundFile) {
    while(!finished.load()) {
        if(playAlert.load()) {
            std::string command = "mpg123 " + soundFile;
            system(command.c_str());
        }
        // Sleep briefly to prevent tight loop when not playing
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
}

void eyeDetectionThread(cv::CascadeClassifier& eye_cascade, const int framesThreshold, cv::VideoCapture& cap) {
    int framesWithoutEyes = 5;
    cv::Mat frame;

    while (!finished.load()) {
        cap >> frame;
        if (frame.empty()) {
            finished.store(true);
            break;
        }

        cv::Mat gray;
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
        cv::equalizeHist(gray, gray);

        std::vector<cv::Rect> eyes;
        eye_cascade.detectMultiScale(gray, eyes, 1.25, 12);

        for(const auto& eye : eyes) {
            cv::rectangle(frame, eye, cv::Scalar(0, 255, 0), 2);
        }

        if (eyes.empty()) {
            framesWithoutEyes++;
            if (framesWithoutEyes >= framesThreshold) {
                playAlert.store(true);
            }
        } else {
            framesWithoutEyes = 0;
            playAlert.store(false);
        }

        cv::imshow("Driver Drowsiness Detection", frame);
        if (cv::waitKey(1) == 'q') {
            finished.store(true);
            break;
        }
    }
}

int main() {
    cv::VideoCapture cap(0);
    if (!cap.isOpened()) {
        std::cerr << "Error opening video stream" << std::endl;
        return -1;
    }

    cap.set(cv::CAP_PROP_FRAME_WIDTH, 640);
    cap.set(cv::CAP_PROP_FRAME_HEIGHT, 480);

    cv::CascadeClassifier eye_cascade;
    if (!eye_cascade.load("/usr/share/opencv4/haarcascades/haarcascade_eye.xml")) {
        std::cerr << "Error loading eye cascade" << std::endl;
        return -1;
    }

    const int framesThreshold = 10;
    std::thread soundThread(playSound, "./alarm.mp3");
    std::thread eyeDetectionWorker(eyeDetectionThread, std::ref(eye_cascade), framesThreshold, std::ref(cap));

    if (eyeDetectionWorker.joinable()) {
        eyeDetectionWorker.join();
    }
    if (soundThread.joinable()) {
        soundThread.join();
    }

    cap.release();
    cv::destroyAllWindows();
    return 0;
}
